<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorials &mdash; simAIRR  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/my_theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Configuration manual" href="configuration.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            simAIRR
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Tool overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration manual</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#running-simairr">Running simAIRR</a></li>
<li class="toctree-l2"><a class="reference internal" href="#configuration-file-format">Configuration file format</a></li>
<li class="toctree-l2"><a class="reference internal" href="#signal-implantation-mode-without-much-customisation">Signal implantation mode (without much customisation)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#signal-implantation-mode-more-customisation">Signal implantation mode (more customisation)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#signal-feasibility-assessment-mode">Signal feasibility assessment mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="#baseline-repertoires-generation">Baseline repertoires generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#public-component-correction">Public component correction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#querying-sequences-enriched-for-k-mer-like-patterns">Querying sequences enriched for k-mer like patterns</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#generation-of-reference-sequences-to-query-against">Generation of reference sequences to query against</a></li>
<li class="toctree-l3"><a class="reference internal" href="#pattern-matching-recipe-using-bionumpy">Pattern matching recipe using bionumpy</a></li>
<li class="toctree-l3"><a class="reference internal" href="#presence-of-multiple-motifs-in-sequences">Presence of multiple motifs in sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="#recipe-using-grep">Recipe using grep</a></li>
<li class="toctree-l3"><a class="reference internal" href="#conditioning-on-gene-usage">Conditioning on gene usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#wall-time-estimate-for-pattern-matching">Wall time estimate for pattern matching</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#file-format-of-ground-truth-signal-sequences">File format of ground truth signal sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pgen-count-mapping-file-format">pgen_count_mapping file format</a></li>
<li class="toctree-l2"><a class="reference internal" href="#custom-models-of-realistic-receptor-sharing-for-different-airr-loci">Custom models of realistic receptor sharing for different AIRR loci</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">simAIRR</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorials</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/tutorials.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="tutorials">
<span id="user-guide"></span><h1>Tutorials<a class="headerlink" href="#tutorials" title="Permalink to this heading"></a></h1>
<div class="section" id="running-simairr">
<h2>Running simAIRR<a class="headerlink" href="#running-simairr" title="Permalink to this heading"></a></h2>
<p>simAIRR can be run through a command-line interface with a single user-supplied parameter - the path to the configuration specification file. Example below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>sim_airr<span class="w"> </span>-i<span class="w"> </span>&lt;path_to_specification_file&gt;
</pre></div>
</div>
</div>
<div class="section" id="configuration-file-format">
<h2>Configuration file format<a class="headerlink" href="#configuration-file-format" title="Permalink to this heading"></a></h2>
<p>The configuration for simAIRR’s simulations are to be supplied through a flat config file in <a class="reference external" href="https://en.wikipedia.org/wiki/YAML">YAML format</a>. The snippets below show examples of the configuration for different usage modes of simAIRR. See <a class="reference internal" href="configuration.html#configuration-table"><span class="std std-ref">simAIRR arguments</span></a> for description and default settings of parameters.</p>
</div>
<div class="section" id="signal-implantation-mode-without-much-customisation">
<h2>Signal implantation mode (without much customisation)<a class="headerlink" href="#signal-implantation-mode-without-much-customisation" title="Permalink to this heading"></a></h2>
<p>In the example specification below, since no values were supplied for the non-required arguments, default settings will be used for some parameters (that are not shown in the snippet below). See <a class="reference internal" href="configuration.html#configuration-table"><span class="std std-ref">simAIRR arguments</span></a> for default settings of other parameters.</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">mode</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">signal_feasibility_assessment</span>
<span class="nt">olga_model</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="nt">n_repertoires</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span>
<span class="nt">n_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span>
<span class="nt">n_threads</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<span class="nt">signal_sequences_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/ground_truth_signal_sequences.tsv</span>
<span class="nt">phenotype_burden</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3</span>
</pre></div>
</div>
</div>
<div class="section" id="signal-implantation-mode-more-customisation">
<h2>Signal implantation mode (more customisation)<a class="headerlink" href="#signal-implantation-mode-more-customisation" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>See acceptable file formats of signal sequences file <a class="reference internal" href="#signal-seq-file-format"><span class="std std-ref">File format of ground truth signal sequences</span></a></p></li>
</ul>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">mode</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">signal_implantation</span>
<span class="nt">olga_model</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="nt">output_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/output/directory</span>
<span class="nt">n_repertoires</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span>
<span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1234</span>
<span class="nt">store_intermediate_files</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="nt">n_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span>
<span class="nt">n_threads</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<span class="nt">public_seq_proportion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.08</span><span class="w"> </span><span class="c1"># 8% of the total unique sequences will be shared across repertoires</span>
<span class="nt">public_seq_pgen_count_mapping_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/public_seq_pgen_count_mapping.tsv</span>
<span class="nt">signal_pgen_count_mapping_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/signal_pgen_count_mapping.tsv</span>
<span class="nt">signal_sequences_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/ground_truth_signal_sequences.tsv</span>
<span class="nt">positive_label_rate</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.3</span><span class="w"> </span><span class="c1"># 30% of the total repertoires will receive signal implantation</span>
<span class="nt">phenotype_burden</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3</span>
<span class="nt">phenotype_pool_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span><span class="w"> </span><span class="c1"># use only any 100 of the total supplied sequences as a signal superset</span>
<span class="nt">allow_closer_phenotype_burden</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to allow signal implantation if the feasible phenotype burden is closer to the user-desired phenotype burden (although it is not exact match)</span>
<span class="nt">export_nt</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to export nucleotide sequences in the simulated repertoire files</span>
</pre></div>
</div>
</div>
<div class="section" id="signal-feasibility-assessment-mode">
<h2>Signal feasibility assessment mode<a class="headerlink" href="#signal-feasibility-assessment-mode" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>It is possible that user-supplied signal sequences cannot be sufficient to meet the user-desired specifications of implantation statistics including desired phenotype burden, number of positive class-labeled repertoires and so on based on realistic levels of population incidence given the generation probability.</p></li>
<li><p>In such a case, <cite>signal_implantation</cite> mode will not succeed in generating simulated repertoires. Rather, guidance will be provided on the signal implantation statistics (what is realistically possible) that will help the user to either tune the simulation parameters or supply a different set of signal sequences or both.</p></li>
<li><p>To help with this process of fine-tuning a set of signal sequences file and parameters of simulation, <cite>signal_feasibility_assessment</cite> mode coule be used. All the parameters of <cite>signal_implantation</cite> mode and <cite>signal_feasibility_assessment</cite> mode are identical. The only difference is that in <cite>signal_implantation</cite> mode, if implantation was deemed feasible, it will proceed with expensive computation steps of baseline repertoire generation and public component correction, whereas the <cite>signal_feasibility_assessment</cite> mode will stop even if the implantation was deemed feasible.</p></li>
</ul>
</div>
<div class="section" id="baseline-repertoires-generation">
<span id="id1"></span><h2>Baseline repertoires generation<a class="headerlink" href="#baseline-repertoires-generation" title="Permalink to this heading"></a></h2>
<p>Since the <cite>signal_implantation</cite> mode involves a sequence of steps that also involves baseline repertoires generation, this functionality is also made available to be run in a separate mode. Although generation of baseline repertoires can be accomplished with a few lines of code around existing tools, the parallelised version of this functionality implemented in <cite>baseline_repertoire_generation</cite> mode may turn out to be useful.</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">mode</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">baseline_repertoire_generation</span>
<span class="nt">olga_model</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="nt">output_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/output/directory</span>
<span class="nt">n_repertoires</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">20</span>
<span class="nt">n_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span>
<span class="nt">n_threads</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
</pre></div>
</div>
</div>
<div class="section" id="public-component-correction">
<h2>Public component correction<a class="headerlink" href="#public-component-correction" title="Permalink to this heading"></a></h2>
<p>When synthetic AIRR datasets are generated using sampling from know V(D)J recombination models using existing tools, the resulting repertoires represent naive repertoires that have not experienced any antigen events. Thus, the proportion of public (shared) sequences in such AIRR datasets will be lower than what is observed in experimental AIRR datasets of antigen-experienced repertoires. To match real-world experimental datasets in terms of public sequences, simAIRR’s workflows include a public component correction step, where a fraction of the total unique sequences in the synthetic AIRR dataset (<cite>public_seq_proportion</cite>) will be forced to be shared across repertoires. The sharing pattern will be determined based on empirically learnt relation between generation probability and population incidence of sequences. With sampling from known V(D)J recombination models, one cannot exclude the possibility of observing same sequence twice; <cite>public_component_correction</cite> mode filters out duplicate sequences before making the sequences public.</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">mode</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">baseline_repertoire_generation</span>
<span class="nt">olga_model</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="nt">output_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/output/directory</span>
<span class="nt">n_repertoires</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">20</span>
<span class="nt">n_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span>
<span class="nt">n_threads</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<span class="nt">public_seq_proportion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.12</span><span class="w"> </span><span class="c1"># 12% of the total unique sequences will be shared across repertoires. Default is 10% if this argument is not supplied.</span>
<span class="nt">public_seq_pgen_count_mapping_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/public_seq_pgen_count_mapping.tsv</span><span class="w"> </span><span class="c1"># default is a real-world experimental dataset calibrated mapping that is included with simAIRR</span>
</pre></div>
</div>
</div>
<div class="section" id="querying-sequences-enriched-for-k-mer-like-patterns">
<h2>Querying sequences enriched for k-mer like patterns<a class="headerlink" href="#querying-sequences-enriched-for-k-mer-like-patterns" title="Permalink to this heading"></a></h2>
<p>Using simAIRR and the new <a class="reference external" href="https://github.com/bionumpy/bionumpy">bionumpy</a> library, one could query and retrieve sequences enriched for k-mer like patterns very easily with just a few lines of Python code. Below, we provide a simple Python recipe.</p>
<div class="section" id="generation-of-reference-sequences-to-query-against">
<h3>Generation of reference sequences to query against<a class="headerlink" href="#generation-of-reference-sequences-to-query-against" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>First, use the simAIRR’s <a class="reference internal" href="#baseline-repertoires-generation"><span class="std std-ref">Baseline repertoires generation</span></a> mode to generate a large number of reference sequences to query against. For instance, in the code chunk below, we generate a single file with ten million sequences.</p></li>
</ul>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">mode</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">baseline_repertoire_generation</span>
<span class="nt">olga_model</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="nt">output_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/output/directory</span>
<span class="nt">n_repertoires</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">n_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10000000</span>
<span class="nt">n_threads</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
</pre></div>
</div>
<p>Note that instead of generating 10 million sequences in a single file, if multiple processes are used using <cite>n_threads</cite> argument (if more CPUs are available), and the n_sequences are split across multiple repertoires using <cite>n_threads</cite> argument, the reference sequence generation will be much quicker. For instance, the following configuration will generate 10 million sequences across 40 files in less than 2 minutes wall time.</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">mode</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">baseline_repertoire_generation</span>
<span class="nt">olga_model</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="nt">output_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/path/to/output/directory</span>
<span class="nt">n_repertoires</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">40</span>
<span class="nt">n_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">250000</span>
<span class="nt">n_threads</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">40</span>
</pre></div>
</div>
<p>To combine them into a single file:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat<span class="w"> </span>/path/to/output/directory/*<span class="w"> </span>&gt;<span class="w"> </span>reference_sequences.txt
</pre></div>
</div>
<ul class="simple">
<li><p>Alternatively, one could use the reference sequences of their choice to query the sequences for k-mer like patterns. The code chunks above is just one example to get a large number of sequences to make queries.</p></li>
<li><p>Let us assume that the ten million sequences that were generated in previous step are stored in a file named <code class="docutils literal notranslate"><span class="pre">reference_sequences.txt</span></code>.</p></li>
</ul>
</div>
<div class="section" id="pattern-matching-recipe-using-bionumpy">
<h3>Pattern matching recipe using bionumpy<a class="headerlink" href="#pattern-matching-recipe-using-bionumpy" title="Permalink to this heading"></a></h3>
<ul>
<li><p>Next, we import relevant functionalities from the bionumpy library to read in the <code class="docutils literal notranslate"><span class="pre">reference_sequences.txt</span></code>, create a lookup index and querying the sequences.</p></li>
<li><p>bionumpy works with popular file formats of biological sequence data. Since bionumpy does not know the file format of our <code class="docutils literal notranslate"><span class="pre">reference_sequences.txt</span></code>, we first need to tell bionumpy that the first two fields of this custom file format are a dna sequence and an amino acid sequence. The code chunk below shows how to do this. This is routine code and can just be copied as shown below.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.io.delimited_buffers</span> <span class="kn">import</span> <span class="n">DelimitedBuffer</span><span class="p">,</span> <span class="n">get_bufferclass_for_datatype</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.bnpdataclass</span> <span class="kn">import</span> <span class="n">bnpdataclass</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.sequence.indexing.wildcard_index</span> <span class="kn">import</span> <span class="n">WildCardLookup</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nd">@bnpdataclass</span>
<span class="gp">... </span><span class="k">class</span> <span class="nc">Olga</span><span class="p">:</span>
<span class="gp">... </span>  <span class="n">dna</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span>
<span class="gp">... </span>  <span class="n">amino_acid</span><span class="p">:</span> <span class="n">bnp</span><span class="o">.</span><span class="n">AminoAcidEncoding</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">class</span> <span class="nc">OlgaBuffer</span><span class="p">(</span><span class="n">DelimitedBuffer</span><span class="p">):</span>
<span class="gp">... </span>   <span class="n">dataclass</span> <span class="o">=</span> <span class="n">Olga</span>
</pre></div>
</div>
</li>
<li><p>Using the newly created buffer type, we read-in the files with that file format.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">olga_sequence_data</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;reference_sequences.txt&quot;</span><span class="p">,</span> <span class="n">buffer_type</span><span class="o">=</span><span class="n">OlgaBuffer</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">olga_sequence_data</span>
<span class="go">Olga with 10000000 entries</span>
<span class="go">                      dna               amino_acid</span>
<span class="go">  TGCGCCACCTGGGGGGACGA...               CATWGDEQYF</span>
<span class="go">  TGTGCCAGCTCACCTACGAA...         CASSPTNSGSNYGYTF</span>
<span class="go">  TGCGGGCCCGTAATGAACAC...              CGPVMNTEAFF</span>
<span class="go">  TGTGCCAGCAGTGAAGCGCG...         CASSEARPARMYGYTF</span>
<span class="go">  TGTGCCAGCAGTAGTGGGAC...          CASSSGTGPDQPQHF</span>
<span class="go">  TGTGCCAGCAACCTAGCGGG...          CASNLAGKNTGELFF</span>
<span class="go">  TGTGCCAGCAGCCAACCGGG...         CASSQPGGSGNYGYTF</span>
<span class="go">  TGCGCCAGCAGCCGCGGCCT...           CASSRGLREETQYF</span>
<span class="go">  TGTGCCAGCAGCCAAGTCTC...        CASSQVSRQDSSYEQYF</span>
<span class="go">  TGTGCCAGCAGGCCGGGACA...     CASRPGQGAPGWEDNYGYTF</span>
</pre></div>
</div>
</li>
<li><p>We create a wildcard lookup table on the amino acid sequences of the file that we read-in.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aminoacid_wildcard_lookup</span> <span class="o">=</span> <span class="n">WildCardLookup</span><span class="o">.</span><span class="n">create_lookup</span><span class="p">(</span><span class="n">olga_sequence_data</span><span class="o">.</span><span class="n">amino_acid</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aminoacid_wildcard_lookup</span>
<span class="go">Lookup on WildcardIndex of 10000000 sequences</span>
</pre></div>
</div>
</li>
<li><p>We then retrieve all the sequences that contain a pattern like “RG.”, where the “.” indicates a wildcard character.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rg_wildcard_sequences</span> <span class="o">=</span> <span class="n">aminoacid_wildcard_lookup</span><span class="o">.</span><span class="n">get_sequences</span><span class="p">(</span><span class="s2">&quot;RG.&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg_wildcard_sequences</span>
<span class="go">encoded_ragged_array([&#39;CASSRGLREETQYF&#39;,</span>
<span class="go">                      &#39;CASGCRGTSGGASLDEQFF&#39;,</span>
<span class="go">                      &#39;CATSDLGVRRGALIATNEKL...&#39;,</span>
<span class="go">                      &#39;CATRRGYGYTF&#39;,</span>
<span class="go">                      &#39;CATRGAKRIDEQYF&#39;,</span>
<span class="go">                      &#39;CASSLRGQGLLRGNQPQHF&#39;,</span>
<span class="go">                      &#39;CASSFSCLRGESSYNEQFF&#39;,</span>
<span class="go">                      &#39;CASRGLFPQPQHF&#39;,</span>
<span class="go">                      &#39;CASSGCRGGNTEAFF&#39;,</span>
<span class="go">                      &#39;CASSEADRGRKAFF&#39;,</span>
<span class="go">                      &#39;CASRVASRGRDKQPQHF&#39;], AlphabetEncoding(&#39;ACDEFGHIKLMNPQRSTVWY&#39;))</span>
</pre></div>
</div>
</li>
</ul>
</div>
<div class="section" id="presence-of-multiple-motifs-in-sequences">
<h3>Presence of multiple motifs in sequences<a class="headerlink" href="#presence-of-multiple-motifs-in-sequences" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>One may be interested in multiple patterns in sequences. For instance, let us assume that one is interested in sequences that contain either of the following pattern: <code class="docutils literal notranslate"><span class="pre">GG**GG</span></code> or <code class="docutils literal notranslate"><span class="pre">GG**YG</span></code>. One can consider this as two cases:</p></li>
</ul>
<ol class="arabic simple">
<li><p>One, where the user is interested in mining those sequences that contain patterns <code class="docutils literal notranslate"><span class="pre">GG</span></code> and either of another <code class="docutils literal notranslate"><span class="pre">GG</span></code> or <code class="docutils literal notranslate"><span class="pre">YG</span></code> that are always two residues apart, where the two residues can be any residues.</p></li>
<li><p>Alternatively, mining for the above patterns will also result in sequences that always contain a <code class="docutils literal notranslate"><span class="pre">GG</span></code> in addition to another <code class="docutils literal notranslate"><span class="pre">GG</span></code> or <code class="docutils literal notranslate"><span class="pre">YG</span></code>.</p></li>
</ol>
<p>In either case, the following code block will mine the sequences of interest:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gg_gg_wildcard_sequences</span> <span class="o">=</span> <span class="n">aminoacid_wildcard_lookup</span><span class="o">.</span><span class="n">get_sequences</span><span class="p">(</span><span class="s2">&quot;GG..GG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gg_gg_wildcard_sequences</span>
<span class="go">encoded_ragged_array([&#39;CASSVKPSGGVAGGETQYF&#39;,</span>
<span class="go">                      &#39;CSARPGGGEGGENQPQHF&#39;,</span>
<span class="go">                      &#39;CASGGRAGGVWAYEQYF&#39;,</span>
<span class="go">                      &#39;CASSTEGGLAGGLTYNEQFF&#39;,</span>
<span class="go">                        ...</span>
<span class="go">                      &#39;CASSGGPEGGFSSTDTQYF&#39;], AlphabetEncoding(&#39;ACDEFGHIKLMNPQRSTVWY&#39;))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">gg_gg_wildcard_sequences</span><span class="p">)</span>
<span class="go">14968</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gg_yg_wildcard_sequences</span> <span class="o">=</span> <span class="n">aminoacid_wildcard_lookup</span><span class="o">.</span><span class="n">get_sequences</span><span class="p">(</span><span class="s2">&quot;GG..YG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">gg_yg_wildcard_sequences</span><span class="p">)</span>
<span class="go">28070</span>
</pre></div>
</div>
<p>If the user is interested in filtering the search results above based on further criteria like gene usage or for containing additional motifs, the resulting sequence set can be considered a proxy for complex signal based on multiple criteria.</p>
</div>
<div class="section" id="recipe-using-grep">
<h3>Recipe using grep<a class="headerlink" href="#recipe-using-grep" title="Permalink to this heading"></a></h3>
<p>Alternatively, one could use the unix <code class="docutils literal notranslate"><span class="pre">grep</span></code> function instead. If the user wants to query multiple patterns from the reference sequences, all the patterns could be listed in one file as shown below. In the example below, three patterns are listed in the file named <code class="docutils literal notranslate"><span class="pre">query_patterns.txt</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat<span class="w"> </span>query_patterns.txt
<span class="go">WKDY</span>
<span class="go">ERFY</span>
<span class="go">YREV</span>
</pre></div>
</div>
<ul class="simple">
<li><p>We then query the reference sequences using the patterns from the file <code class="docutils literal notranslate"><span class="pre">query_patterns.txt</span></code> and write the output to <code class="docutils literal notranslate"><span class="pre">search_results.txt</span></code>.</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>grep<span class="w"> </span>-f<span class="w"> </span>query_patterns.txt<span class="w"> </span>reference_sequences_file.txt<span class="w"> </span>&gt;<span class="w"> </span>search_results.txt
</pre></div>
</div>
<ul class="simple">
<li><p>We retrieved a total of 705 sequences from the 10 million sequences that contained any of the three patterns listed in <code class="docutils literal notranslate"><span class="pre">query_patterns.txt</span></code>.</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>wc<span class="w"> </span>-l<span class="w"> </span>search_results.txt
<span class="go">705</span>
</pre></div>
</div>
</div>
<div class="section" id="conditioning-on-gene-usage">
<h3>Conditioning on gene usage<a class="headerlink" href="#conditioning-on-gene-usage" title="Permalink to this heading"></a></h3>
<p>If one is interested in using/testing sequences that match only a particular V or J gene, the <code class="docutils literal notranslate"><span class="pre">search_results.txt</span></code> could further be filtered to match the specific genes of interest. As an example, if the user is interested only in those sequences where TRBV5 or TRBV27 genes are used, one could further filter the <code class="docutils literal notranslate"><span class="pre">search_results.txt</span></code> as shown below:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>grep<span class="w"> </span><span class="s2">&quot;TRBV27\|TRBV5&quot;</span><span class="w"> </span>search_results.txt<span class="w"> </span>&gt;<span class="w"> </span>search_results_trbv5_trbv27.txt
<span class="gp">$ </span>wc<span class="w"> </span>-l<span class="w"> </span>search_results_trbv5_trbv27.txt
<span class="go">114</span>
</pre></div>
</div>
</div>
<div class="section" id="wall-time-estimate-for-pattern-matching">
<h3>Wall time estimate for pattern matching<a class="headerlink" href="#wall-time-estimate-for-pattern-matching" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>In the examples that we have shown above, generation of 10 million sequences on 40 CPUs took less than 1 minute wall time and the query of the reference sequences using either bionumpy or unix grep took less than 1 minute wall time when we queried three patterns. The query time can increase with increase in the number of patterns, but for AIRR benchmarking datasets-relevant number of patterns, the query time could still be reasonably efficient.</p></li>
</ul>
</div>
</div>
<div class="section" id="file-format-of-ground-truth-signal-sequences">
<span id="signal-seq-file-format"></span><h2>File format of ground truth signal sequences<a class="headerlink" href="#file-format-of-ground-truth-signal-sequences" title="Permalink to this heading"></a></h2>
<p>The ground truth signal sequences file is expected to be a tab-delimited file with at least 3 columns and without a header line. See below for file formats that are accepted.</p>
<p>Supplying nucleotide sequences is optional. If nucleotide sequences are not supplied,
nucleotide sequences are not exported in the simulated repertoires by default. This behavior can be changed with <cite>export_nt</cite> parameter.</p>
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">When including nucleotide sequences</span><a class="headerlink" href="#id2" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>TGTGCCAGCAGTTTATCGCCGGGACTGGCCTACGAGCAGTACTTC</p></td>
<td><p>CASSLSPGLAYEQYF</p></td>
<td><p>TRBV27</p></td>
<td><p>TRBJ2-7</p></td>
</tr>
<tr class="row-even"><td><p>TGTGCCAGCAAAGTCAGAATTGCTGCAACTAATGAAAAACTGTTTTTT</p></td>
<td><p>CASKVRIAATNEKLFF</p></td>
<td><p>TRBV5-6</p></td>
<td><p>TRBJ1-4</p></td>
</tr>
<tr class="row-odd"><td><p>TGCAGTGCCGACTCCAAGAACAGAGGAGCGGGGGGGGAGGCAAGCTCCTACGAGCAGTACTTC</p></td>
<td><p>CSADSKNRGAGGEASSYEQYF</p></td>
<td><p>TRBV20-1</p></td>
<td><p>TRBJ2-7</p></td>
</tr>
<tr class="row-even"><td><p>TGTGCCAGCATCGGTGGCGGGACTAGTCTCTCCTACAATGAGCAGTTCTTC</p></td>
<td><p>CASIGGGTSLSYNEQFF</p></td>
<td><p>TRBV7-9</p></td>
<td><p>TRBJ2-1</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id3">
<caption><span class="caption-text">When not including nucleotide sequences</span><a class="headerlink" href="#id3" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>.</p></td>
<td><p>CASSLSPGLAYEQYF</p></td>
<td><p>TRBV27</p></td>
<td><p>TRBJ2-7</p></td>
</tr>
<tr class="row-even"><td><p>.</p></td>
<td><p>CASKVRIAATNEKLFF</p></td>
<td><p>TRBV5-6</p></td>
<td><p>TRBJ1-4</p></td>
</tr>
<tr class="row-odd"><td><p>.</p></td>
<td><p>CSADSKNRGAGGEASSYEQYF</p></td>
<td><p>TRBV20-1</p></td>
<td><p>TRBJ2-7</p></td>
</tr>
<tr class="row-even"><td><p>.</p></td>
<td><p>CASIGGGTSLSYNEQFF</p></td>
<td><p>TRBV7-9</p></td>
<td><p>TRBJ2-1</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id4">
<caption><span class="caption-text">When not including nucleotide sequences</span><a class="headerlink" href="#id4" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>NA</p></td>
<td><p>CASSLSPGLAYEQYF</p></td>
<td><p>TRBV27</p></td>
<td><p>TRBJ2-7</p></td>
</tr>
<tr class="row-even"><td><p>NA</p></td>
<td><p>CASKVRIAATNEKLFF</p></td>
<td><p>TRBV5-6</p></td>
<td><p>TRBJ1-4</p></td>
</tr>
<tr class="row-odd"><td><p>NA</p></td>
<td><p>CSADSKNRGAGGEASSYEQYF</p></td>
<td><p>TRBV20-1</p></td>
<td><p>TRBJ2-7</p></td>
</tr>
<tr class="row-even"><td><p>NA</p></td>
<td><p>CASIGGGTSLSYNEQFF</p></td>
<td><p>TRBV7-9</p></td>
<td><p>TRBJ2-1</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id5">
<caption><span class="caption-text">When not supplying nucleotide sequences field</span><a class="headerlink" href="#id5" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>CASSLSPGLAYEQYF</p></td>
<td><p>TRBV27</p></td>
<td><p>TRBJ2-7</p></td>
</tr>
<tr class="row-even"><td><p>CASKVRIAATNEKLFF</p></td>
<td><p>TRBV5-6</p></td>
<td><p>TRBJ1-4</p></td>
</tr>
<tr class="row-odd"><td><p>CSADSKNRGAGGEASSYEQYF</p></td>
<td><p>TRBV20-1</p></td>
<td><p>TRBJ2-7</p></td>
</tr>
<tr class="row-even"><td><p>CASIGGGTSLSYNEQFF</p></td>
<td><p>TRBV7-9</p></td>
<td><p>TRBJ2-1</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="pgen-count-mapping-file-format">
<h2>pgen_count_mapping file format<a class="headerlink" href="#pgen-count-mapping-file-format" title="Permalink to this heading"></a></h2>
<p>The user is not required to supply this file unless the user desires to generate simulated datasets based on a custom model derived from a specific experimental dataset.</p>
<ul class="simple">
<li><p>For both the signal sequences and remaining public sequences, user could supply custom empirical relation between generation probability and population incidence. The file format for both of those files is shown below. The file should be tab-delimited with required fields: <cite>“pgen_left”, “pgen_right”, “sample_size_prop_left”, “sample_size_prop_right”, “prob”</cite>.</p></li>
<li><p>To prepare such files based on a dataset of interest, one should compute empirical probabilities (<cite>prob</cite>) of observing sequences within a range of population incidence levels (<cite>sample_size_prop_left</cite> and <cite>sample_size_prop_right</cite>) for a given range of generation probability (<cite>pgen_left</cite> and <cite>pgen_right</cite>).</p></li>
<li><p>An example of such file is shown below:</p></li>
</ul>
<table class="docutils align-default" id="id6">
<caption><span class="caption-text">pgen_count_mapping file format and example</span><a class="headerlink" href="#id6" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>pgen_left</p></th>
<th class="head"><p>pgen_right</p></th>
<th class="head"><p>sample_size_prop_left</p></th>
<th class="head"><p>sample_size_prop_right</p></th>
<th class="head"><p>prob</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>-0.001</p></td>
<td><p>1e-20</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.8867159551633675</p></td>
</tr>
<tr class="row-odd"><td><p>-0.001</p></td>
<td><p>1e-20</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.04817553064631529</p></td>
</tr>
<tr class="row-even"><td><p>-0.001</p></td>
<td><p>1e-20</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.025280228953016935</p></td>
</tr>
<tr class="row-odd"><td><p>-0.001</p></td>
<td><p>1e-20</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.022776055330312427</p></td>
</tr>
<tr class="row-even"><td><p>-0.001</p></td>
<td><p>1e-20</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.010255187216789887</p></td>
</tr>
<tr class="row-odd"><td><p>-0.001</p></td>
<td><p>1e-20</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.004889100882423086</p></td>
</tr>
<tr class="row-even"><td><p>-0.001</p></td>
<td><p>1e-20</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>0.0019079418077748629</p></td>
</tr>
<tr class="row-odd"><td><p>1e-20</p></td>
<td><p>1e-15</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.9316377432726761</p></td>
</tr>
<tr class="row-even"><td><p>1e-20</p></td>
<td><p>1e-15</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.03456755427527577</p></td>
</tr>
<tr class="row-odd"><td><p>1e-20</p></td>
<td><p>1e-15</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.01352490690648493</p></td>
</tr>
<tr class="row-even"><td><p>1e-20</p></td>
<td><p>1e-15</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.012225110658329236</p></td>
</tr>
<tr class="row-odd"><td><p>1e-20</p></td>
<td><p>1e-15</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.0036183517178388254</p></td>
</tr>
<tr class="row-even"><td><p>1e-20</p></td>
<td><p>1e-15</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.002845499894611115</p></td>
</tr>
<tr class="row-odd"><td><p>1e-20</p></td>
<td><p>1e-15</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>0.0015808332747839528</p></td>
</tr>
<tr class="row-even"><td><p>1e-15</p></td>
<td><p>1e-14</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.8809934274323887</p></td>
</tr>
<tr class="row-odd"><td><p>1e-15</p></td>
<td><p>1e-14</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.0603652623639694</p></td>
</tr>
<tr class="row-even"><td><p>1e-15</p></td>
<td><p>1e-14</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.026640448227561685</p></td>
</tr>
<tr class="row-odd"><td><p>1e-15</p></td>
<td><p>1e-14</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.023273354164421937</p></td>
</tr>
<tr class="row-even"><td><p>1e-15</p></td>
<td><p>1e-14</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.006599504363753905</p></td>
</tr>
<tr class="row-odd"><td><p>1e-15</p></td>
<td><p>1e-14</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.0019663829328736126</p></td>
</tr>
<tr class="row-even"><td><p>1e-15</p></td>
<td><p>1e-14</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>0.0001616205150307079</p></td>
</tr>
<tr class="row-odd"><td><p>1e-14</p></td>
<td><p>1e-13</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.768952812809976</p></td>
</tr>
<tr class="row-even"><td><p>1e-14</p></td>
<td><p>1e-13</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.0895800859666525</p></td>
</tr>
<tr class="row-odd"><td><p>1e-14</p></td>
<td><p>1e-13</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.057413442917198056</p></td>
</tr>
<tr class="row-even"><td><p>1e-14</p></td>
<td><p>1e-13</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.048438902271975816</p></td>
</tr>
<tr class="row-odd"><td><p>1e-14</p></td>
<td><p>1e-13</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.024892541684379575</p></td>
</tr>
<tr class="row-even"><td><p>1e-14</p></td>
<td><p>1e-13</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.010249870105332766</p></td>
</tr>
<tr class="row-odd"><td><p>1e-14</p></td>
<td><p>1e-13</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>0.00047234424448538097</p></td>
</tr>
<tr class="row-even"><td><p>1e-13</p></td>
<td><p>1e-12</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.7716265865063461</p></td>
</tr>
<tr class="row-odd"><td><p>1e-13</p></td>
<td><p>1e-12</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.06808060565575595</p></td>
</tr>
<tr class="row-even"><td><p>1e-13</p></td>
<td><p>1e-12</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.05934090403028279</p></td>
</tr>
<tr class="row-odd"><td><p>1e-13</p></td>
<td><p>1e-12</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.0412769984413271</p></td>
</tr>
<tr class="row-even"><td><p>1e-13</p></td>
<td><p>1e-12</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.03406813627254509</p></td>
</tr>
<tr class="row-odd"><td><p>1e-13</p></td>
<td><p>1e-12</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.02265642395902917</p></td>
</tr>
<tr class="row-even"><td><p>1e-13</p></td>
<td><p>1e-12</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>0.002950345134713872</p></td>
</tr>
<tr class="row-odd"><td><p>1e-12</p></td>
<td><p>1e-11</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.9637811113220949</p></td>
</tr>
<tr class="row-even"><td><p>1e-12</p></td>
<td><p>1e-11</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.013495931528718414</p></td>
</tr>
<tr class="row-odd"><td><p>1e-12</p></td>
<td><p>1e-11</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.007609483019319085</p></td>
</tr>
<tr class="row-even"><td><p>1e-12</p></td>
<td><p>1e-11</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.00584425174589109</p></td>
</tr>
<tr class="row-odd"><td><p>1e-12</p></td>
<td><p>1e-11</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.004486922519709405</p></td>
</tr>
<tr class="row-even"><td><p>1e-12</p></td>
<td><p>1e-11</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.003966495769774458</p></td>
</tr>
<tr class="row-odd"><td><p>1e-12</p></td>
<td><p>1e-11</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>0.0008158040944926191</p></td>
</tr>
<tr class="row-even"><td><p>1e-11</p></td>
<td><p>1e-10</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.9675325841713831</p></td>
</tr>
<tr class="row-odd"><td><p>1e-11</p></td>
<td><p>1e-10</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.02028032266705611</p></td>
</tr>
<tr class="row-even"><td><p>1e-11</p></td>
<td><p>1e-10</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.006322762334892417</p></td>
</tr>
<tr class="row-odd"><td><p>1e-11</p></td>
<td><p>1e-10</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.004386728759275038</p></td>
</tr>
<tr class="row-even"><td><p>1e-11</p></td>
<td><p>1e-10</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.0010441617146431771</p></td>
</tr>
<tr class="row-odd"><td><p>1e-11</p></td>
<td><p>1e-10</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.0003826770681938345</p></td>
</tr>
<tr class="row-even"><td><p>1e-11</p></td>
<td><p>1e-10</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>5.076328455632499e-05</p></td>
</tr>
<tr class="row-odd"><td><p>1e-10</p></td>
<td><p>1e-09</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.8350908568888563</p></td>
</tr>
<tr class="row-even"><td><p>1e-10</p></td>
<td><p>1e-09</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.07976968571914192</p></td>
</tr>
<tr class="row-odd"><td><p>1e-10</p></td>
<td><p>1e-09</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.036914283085982824</p></td>
</tr>
<tr class="row-even"><td><p>1e-10</p></td>
<td><p>1e-09</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.035715530564775576</p></td>
</tr>
<tr class="row-odd"><td><p>1e-10</p></td>
<td><p>1e-09</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.009978644525678853</p></td>
</tr>
<tr class="row-even"><td><p>1e-10</p></td>
<td><p>1e-09</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.0024663183241324764</p></td>
</tr>
<tr class="row-odd"><td><p>1e-10</p></td>
<td><p>1e-09</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>6.468089143204572e-05</p></td>
</tr>
<tr class="row-even"><td><p>1e-09</p></td>
<td><p>1e-08</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.4951827356202455</p></td>
</tr>
<tr class="row-odd"><td><p>1e-09</p></td>
<td><p>1e-08</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.1390661087283864</p></td>
</tr>
<tr class="row-even"><td><p>1e-09</p></td>
<td><p>1e-08</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.12290976272212115</p></td>
</tr>
<tr class="row-odd"><td><p>1e-09</p></td>
<td><p>1e-08</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.08791235702984151</p></td>
</tr>
<tr class="row-even"><td><p>1e-09</p></td>
<td><p>1e-08</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.08482276197758802</p></td>
</tr>
<tr class="row-odd"><td><p>1e-09</p></td>
<td><p>1e-08</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.06275295072463237</p></td>
</tr>
<tr class="row-even"><td><p>1e-09</p></td>
<td><p>1e-08</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>0.00735332319718501</p></td>
</tr>
<tr class="row-odd"><td><p>1e-08</p></td>
<td><p>0.01</p></td>
<td><p>0.019</p></td>
<td><p>0.0512</p></td>
<td><p>0.26781019279511525</p></td>
</tr>
<tr class="row-even"><td><p>1e-08</p></td>
<td><p>0.01</p></td>
<td><p>0.0019299999999999999</p></td>
<td><p>0.00439</p></td>
<td><p>0.17710858570927396</p></td>
</tr>
<tr class="row-odd"><td><p>1e-08</p></td>
<td><p>0.01</p></td>
<td><p>0.0512</p></td>
<td><p>0.943</p></td>
<td><p>0.16489284123474018</p></td>
</tr>
<tr class="row-even"><td><p>1e-08</p></td>
<td><p>0.01</p></td>
<td><p>0.0117</p></td>
<td><p>0.019</p></td>
<td><p>0.14483284651679812</p></td>
</tr>
<tr class="row-odd"><td><p>1e-08</p></td>
<td><p>0.01</p></td>
<td><p>0.00732</p></td>
<td><p>0.0117</p></td>
<td><p>0.12801895233928157</p></td>
</tr>
<tr class="row-even"><td><p>1e-08</p></td>
<td><p>0.01</p></td>
<td><p>0.00439</p></td>
<td><p>0.00586</p></td>
<td><p>0.06279342029019784</p></td>
</tr>
<tr class="row-odd"><td><p>1e-08</p></td>
<td><p>0.01</p></td>
<td><p>0.00586</p></td>
<td><p>0.00732</p></td>
<td><p>0.054543161114593064</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="custom-models-of-realistic-receptor-sharing-for-different-airr-loci">
<h2>Custom models of realistic receptor sharing for different AIRR loci<a class="headerlink" href="#custom-models-of-realistic-receptor-sharing-for-different-airr-loci" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>Note that in simAIRR package, we have provided empirical models for
realistic receptor sequence sharing just for TCRB loci, based on the
public availability of a large experimental dataset. For other AIRR
loci, custom models will be added into future versions of simAIRR
when large-scale experimental data for other loci becomes available.
For now users are encouraged to generate their own custom models of
realistic receptor sharing for other loci (if sufficiently large data
available to them). Note that, based on our analyses, a sample size
of 200 (100 each of cases and controls) approximates well the
realistic receptor sequence sharing.</p></li>
<li><p>Through <a class="reference external" href="https://github.com/KanduriC/usecases_simairr">usecases_simairr</a> python package, we provide console scripts that can be run
through command line to learn custom models for realistic receptor
sequence sharing. Each console script requires some input parameters
from the users. To know which input arguments are required for each
console script, we recommend using the <code class="docutils literal notranslate"><span class="pre">--help</span></code> argument of each
console script. For instance, <code class="docutils literal notranslate"><span class="pre">concat_airr</span> <span class="pre">--help</span></code>.</p></li>
<li><p>Note that the following console scripts are expected to be run in a
sequence.</p></li>
</ul>
<table class="docutils align-default">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Command</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">concat_airr</span></code></p></td>
<td><p>Concatenate the repertoire files into
two separate files that will be needed
by
<a class="reference external" href="https://github.com/uio-bmi/compairr">compAIRR</a>
tool to count the frequencies of each
unique clones</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">run_compairr_seqcounts</span></code></p></td>
<td><p>Run
<a class="reference external" href="https://github.com/uio-bmi/compairr">compAIRR</a>
tool to count the frequencies of each
unique clones.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">compute_pgen_public</span></code></p></td>
<td><p>Compute the generation probabilities
of each sequence using OLGA tool.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">compute_pval</span></code></p></td>
<td><p>Computes p-values as the probability
of observing a public sequence in the
same or higher number of repertoires
as it was observed in for a given
dataset.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">concat_pdata</span></code></p></td>
<td><p>Concatenate the files generated
through previous step to produce one
large file that will be used in the
next step</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">gen_pgen_count_map</span></code></p></td>
<td><p>Generate pgen_count_map files in the
required format desired by simAIRR</p></td>
</tr>
</tbody>
</table>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="configuration.html" class="btn btn-neutral float-left" title="Configuration manual" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Chakravarthi Kanduri.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>